base/array.jl

```diff
@@ -646,6 +646,12 @@ if isdefined(Core, :Compiler)
         return quote
             if $I isa Generator && ($I).f isa Type
                 ($I).f
+            elseif $I isa Tuple
+                if isempty($I) || length($I) > 16
+                    Any
+                else
+                    mapreduce(typeof, typejoin, $I; init=Union{})
+                end
```

Comment:
Ah, good point that inferring `first` on a tuple does not give a very accurate element type! :) Fortunately I think this is dead code for tuples, since tuples have known lengths, and empty tuples have a known element type.